### PR TITLE
Create TimeZoneFormatConfig enum

### PR DESCRIPTION
**Pull request by samchen61661**
Create TimeZoneFormatConfig enum for https://github.com/unicode-org/icu4x/issues/622

### CODE REVIEW

**Comment by nordzilla** [comment]
This looks great so far, and just the kind of enum we want for this. 

As part of this same PR, we will want to also create a new constructor for `TimeZoneFormat` that accepts the enum as an argument. 

For now I'm going to convert this PR to a draft PR and remove the other reviewers.

**Comment by nordzilla** [comment]
@samchen61661 

So there are still a couple things left to do for this PR. 

1) We need to go through and make parts of the API public that should be, and also remove the TODOs associated with the documentation comments. Try grepping for `TODO` and `622`

```
src/zoned_datetime.rs:27:// TODO(#622) link [`TimeZoneFormat`] once it is public.
src/lib.rs:91:// TODO(#622) make the time_zone module public once TimeZoneFormat is public.
src/lib.rs:103:// TODO(#622) re-export TimeZoneFormat once it is made public.
src/format/time_zone.rs:44:// TODO(#622) Make this public once TimeZoneFormat is public.
src/mock/time_zone.rs:11:// TODO(#622) link [`TimeZoneFormat`] appropriately once it is public.
src/time_zone.rs:67:/// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
src/time_zone.rs:92:// TODO(#622) Make TimeZoneFormat public once we have a clean way to provide it options.
src/time_zone.rs:121:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
src/time_zone.rs:137:    // TODO(#622) Make this public once TimeZoneFormat is public.
src/time_zone.rs:277:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
src/time_zone.rs:302:    // TODO(#622) Make this public once TimeZoneFormat is public.
src/time_zone.rs:320:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
src/time_zone.rs:349:    // TODO(#622) Make this public once TimeZoneFormat is public.
src/time_zone.rs:364:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
src/time_zone.rs:389:    // TODO(#622) Make this public once TimeZoneFormat is public.
```

2) We should explore if we can remove the `TimeZoneFormat::try_new()` that takes a pattern, and see if we can logically replace it with the new public constructor that takes a pattern.

**Comment by samchen61661** [comment]
> @samchen61661
> 
> So there are still a couple things left to do for this PR.
> 
> 1. We need to go through and make parts of the API public that should be, and also remove the TODOs associated with the documentation comments. Try grepping for `TODO` and `622`
> 
> ```
> src/zoned_datetime.rs:27:// TODO(#622) link [`TimeZoneFormat`] once it is public.
> src/lib.rs:91:// TODO(#622) make the time_zone module public once TimeZoneFormat is public.
> src/lib.rs:103:// TODO(#622) re-export TimeZoneFormat once it is made public.
> src/format/time_zone.rs:44:// TODO(#622) Make this public once TimeZoneFormat is public.
> src/mock/time_zone.rs:11:// TODO(#622) link [`TimeZoneFormat`] appropriately once it is public.
> src/time_zone.rs:67:/// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
> src/time_zone.rs:92:// TODO(#622) Make TimeZoneFormat public once we have a clean way to provide it options.
> src/time_zone.rs:121:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
> src/time_zone.rs:137:    // TODO(#622) Make this public once TimeZoneFormat is public.
> src/time_zone.rs:277:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
> src/time_zone.rs:302:    // TODO(#622) Make this public once TimeZoneFormat is public.
> src/time_zone.rs:320:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
> src/time_zone.rs:349:    // TODO(#622) Make this public once TimeZoneFormat is public.
> src/time_zone.rs:364:    /// // TODO(#622) Uncomment and update example once TimeZoneFormat is public.
> src/time_zone.rs:389:    // TODO(#622) Make this public once TimeZoneFormat is public.
> ```
> 
> 1. We should explore if we can remove the `TimeZoneFormat::try_new()` that takes a pattern, and see if we can logically replace it with the new public constructor that takes a pattern.

Sure, I will do both in this PR. 

But I have some questions:
1. re: 'We need to go through and make parts of the API public that should be':
It confuses me because `TimeZoneFormat` and `TimeZoneFormatConfig` are marked as public. Are there any things we need to do?

2. I need to have 1 to 1 match between pattern and config. And here are the gaps I found:
- Do we treat `ZZZZ` the same as `OOOO`? They both long localized GMT format, but `ZZZZ` is `GMT-8:00` while `OOOO` is `GMT-08:00`.
- We talked about `V`s during last meeting. And I found pattern supports `VVV` and `VVV`. Do we want to support them for config? If so, we need to add new enums for them.

3. It is more about implementation details. `iso8601_format` in `TimeZoneFormat` takes `IsoFormat` instead of `&IsoFormat`. I wonder why it is designed to take the value instead of the reference? I think I still want to own `IsoFormat` in the `TimeZoneFormatConfig`,

**Comment by nordzilla** [comment]
> But I have some questions:
> 
>     1. re: 'We need to go through and make parts of the API public that should be':
>        It confuses me because `TimeZoneFormat` and `TimeZoneFormatConfig` are marked as public. Are there any things we need to do?

The goal of this PR is to make `TimeZoneFormat` fully `pub` (it is currently `pub(super)`), and provide a new, `pub` `TimeZoneFormatConfig` enum with a `pub` constructor that utilizes the enum, along with making the other `format` functions public (marked with `TODO(#622)`. 


> 
>     2. I need to have 1 to 1 match between pattern and config. And here are the gaps I found:
> 
> 
>     * Do we treat `ZZZZ` the same as `OOOO`? They both long localized GMT format, but `ZZZZ` is `GMT-8:00` while `OOOO` is `GMT-08:00`.

We are not differentiating between `ZZZZ` or `OOOO` in ICU4X. I'm sorry this is not documented more clearly. I will try to improve the clarity of the documentation here. 

The reason that we only have one function called `localized_gmt_format` instead of something like `long_gmt` or `short_gmt` etc. is because we are choosing to always use the localized form of `hourFormat` as specified in the CLDR-JSON data:

https://github.com/unicode-org/cldr-json/blob/83f6144d8aa3d10e932c25924d4c903925b24257/cldr-json/cldr-dates-full/main/ar/timeZoneNames.json#L12

Rather than let developers arbitrarily choose between long or short GMT format, it seems like a better decision to always use the localized form that is specified by the locale. So in ICU4X, both `ZZZZ` and `OOOO` use the same localized GMT format which refers to the locale's specified `hourFormat` entry. 

>     * We talked about `V`s during last meeting. And I found pattern supports `VVV` and `VVV`. Do we want to support them for config? If so, we need to add new enums for them.

I went back and revisited my reasoning for not including `ExemplarCity` in `TimeZoneFormatConfig` as a stand-alone format. It's because we offer `GenericLocation, // e.g. Los Angeles Time` which contains the exemplar city with the localized word(s) for `Time`. This seems like a better option for developers than only the exemplar city by itself. And again, I would prefer to start with fewer options and add them as needed, rather than start with many options and remove them.

>     1. It is more about implementation details. `iso8601_format` in `TimeZoneFormat` takes `IsoFormat` instead of `&IsoFormat`. I wonder why it is designed to take the value instead of the reference? I think I still want to own `IsoFormat` in the `TimeZoneFormatConfig`,

I appreciate you calling this out. I thought that I had made `IsoFormat` a `Copy` type, but apparently I did not. My opinion is that `iso8601_format` should still take `IsoFormat` by value, but we should add `#[derive(Copy)]` above the declaration for the `IsoFormat` enum type.

**Comment by samchen61661** [comment]
Thanks Eric! 

I took a closer look at the code again. I found that `TimeZoneFormat` is dependent on `DateTimeFormat`. For example, here is a snippet of code:
```
        let datetime_format = DateTimeFormat::new(locale, patterns, symbols_data, ordinal_rules);
        let time_zone_format = TimeZoneFormat::try_new(
            datetime_format.locale.clone(),
            datetime_format
                // Only dates have plural variants so we can use any of the patterns for the time segment.
                .patterns
                .clone(),
            zone_provider,
        )?;
```

It feels like I need to create a mapping from pattern in `DateTimeFormat` to config in `TimeZoneFormat` to get rid of the dependency. But because we want to get rid of patterns, I think that's not an ideal solution.

**Comment by nordzilla** [comment]
> Thanks Eric!
> 
> I took a closer look at the code again. I found that `TimeZoneFormat` is dependent on `DateTimeFormat`. For example, here is a snippet of code:
> 
> ```
>         let datetime_format = DateTimeFormat::new(locale, patterns, symbols_data, ordinal_rules);
>         let time_zone_format = TimeZoneFormat::try_new(
>             datetime_format.locale.clone(),
>             datetime_format
>                 // Only dates have plural variants so we can use any of the patterns for the time segment.
>                 .patterns
>                 .clone(),
>             zone_provider,
>         )?;
> ```
> 
> It feels like I need to create a mapping from pattern in `DateTimeFormat` to config in `TimeZoneFormat` to get rid of the dependency. But because we want to get rid of patterns, I think that's not an ideal solution.

@samchen61661 


For now, this is why we're creating two constructors: the pre-existing `try_new()` and our new constructor that takes a config object. 

For this PR, it would be fine to keep the `try_new()` constructor as `pub(super)`, which gets used only by `ZonedDateTimeFormat` and expose our new public constructor. 

It would be great if we could find a way to consolidate and eventually remove `try_new()`, finding a way that `ZoneDateTimeFormat` could have a mapping from its patterns to the config items. 

This is what I meant when I said: 

> We should explore if we can remove the `TimeZoneFormat::try_new()` that takes a pattern, and see if we can logically replace it with the new public constructor that takes a pattern.

But after thinking about it, since we're not including the `VVV` pattern in the public API, it won't be easy to consolidate this. But we can always look at this further in a follow-up PR.

For now, I think the best option would be to keep `try_new()` as `pub(super)` and just remove its TODO comment, only exposing our new constructor as fully `pub`.

**Comment by samchen61661** [comment]
> > Thanks Eric!
> > I took a closer look at the code again. I found that `TimeZoneFormat` is dependent on `DateTimeFormat`. For example, here is a snippet of code:
> > ```
> >         let datetime_format = DateTimeFormat::new(locale, patterns, symbols_data, ordinal_rules);
> >         let time_zone_format = TimeZoneFormat::try_new(
> >             datetime_format.locale.clone(),
> >             datetime_format
> >                 // Only dates have plural variants so we can use any of the patterns for the time segment.
> >                 .patterns
> >                 .clone(),
> >             zone_provider,
> >         )?;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It feels like I need to create a mapping from pattern in `DateTimeFormat` to config in `TimeZoneFormat` to get rid of the dependency. But because we want to get rid of patterns, I think that's not an ideal solution.
> 
> @samchen61661
> 
> For now, this is why we're creating two constructors: the pre-existing `try_new()` and our new constructor that takes a config object.
> 
> For this PR, it would be fine to keep the `try_new()` constructor as `pub(super)`, which gets used only by `ZonedDateTimeFormat` and only publicly expose our new constructor.
> 
> It would be great if we could find a way to consolidate and eventually remove `try_new()`, finding a way that `ZoneDateTimeFormat` could have a mapping from its patterns to the config items.
> 
> This is what I meant when I said:
> 
> > We should explore if we can remove the `TimeZoneFormat::try_new()` that takes a pattern, and see if we can logically replace it with the new public constructor that takes a pattern.
> 
> But after thinking about it further, since we're not including the `VVV` pattern in the public API, it won't be easy to consolidate this. But we can always look at this further in a follow-up PR.
> 
> For now, I think the best option would be to keep `try_new()` as `pub(super)` and just remove its TODO comment, only exposing our new constructor as fully `pub`.

Sounds good!

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/time_zone.rs:111]: > issue

Sorry I wasn't clear about this. Only the public API (function calls) should become fully `pub`. 

We don't want these data members to be fully exposed. They can go back to being `pub(super)`.

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/time_zone.rs:828]: > praise

Good job adding the `IsoMinutes` and `IsoSeconds` here too. I didn't remember or think about adding those when I was drafting this.

**Review by nordzilla — CHANGES_REQUESTED** [review]
Looking great so far!

There are just a couple changes to make in the review comments, and we also need to add tests for this new functionality before we can land this. 

I'm happy to sync up on a call about testing and walk you through how our test suites work in more detail if that would be helpful.

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/format/time_zone.rs:69]: > suggestion

Now that `TimeZoneFormatConfig` is a `Copy` type, we can remove the `as_ref()` here, because the ownership won't be moved anymore.

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/format/time_zone.rs:106]: > suggestion

Once we remove the `as_ref()` in the comment above, we can remove all of these `.clone()` calls, since everything will be trivially copied.

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [reply components/datetime/src/format/time_zone.rs:69]: > suggestion

Instead of using `unwrap()` here, let's use `expect()` which allows you to also provide a message if this fails. 

We could say something like 
```rust
.expect("TimeZoneFormat::write_config() expected Some(config) but got None")
```

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/format/time_zone.rs:44]: > suggestion

We used to have only `write_pattern()`, which was `pub(crate)` as the top-level abstraction under the public API.

But now we have:
```
Public API
└── write_zone()
    ├── write_pattern()
    └── write_config()
```

So we should be able to leave `write_zone()` as the only `pub(crate)` entrypoint, and make both `write_pattern()` and `write_config()` fully private (i.e. no public specifier at all: only `fn`).

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/format/time_zone.rs:126]: > suggestion 

Same here, we should use `expect()` instead of `unwrap()`

```rust
.expect("TimeZoneFormat::write_pattern() expected Some(patterns) but got None")
```

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/time_zone.rs:180]: > suggestion

Let's use `expect` instead of `unwrap` here as well.

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/time_zone.rs:274]: > nitpick

We should probably rename this to `try_from_config()`. This is a more common naming convention style in Rust.

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/time_zone.rs:822]: @sffc 

Are we okay with adding `Serialize` and `Deserialize` derives here? 

The only reason they exist is for testing. Do you know if these will be removed (essentially skipped) by dead code elimination?

Or should we add equivalent types only in the test module and implement `From<TestType> for ThisType`?

File: components/datetime/tests/datetime.rs
Comment (nordzilla) [inline components/datetime/tests/datetime.rs:282]: > question

Was this supposed to be `tzf` for `TimeZoneFormat`?

**Review by nordzilla — CHANGES_REQUESTED** [review]
Looks really great @samchen61661!

Just a few minor changes, and I'll also add Shane back on as a reviewer.

File: components/datetime/tests/datetime.rs
Comment (nordzilla) [inline components/datetime/tests/datetime.rs:284]: > suggestion

This is a `Copy` type now, so we no longer need an explicit `.clone()`.

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/format/time_zone.rs:56]: > issue

Whoops, I meant to flag this on my previous pass, but it looks like I accidentally skipped it.

This should be `unreachable!()` rather than `todo!()`. 

And we should also be more specific here about what was populated. 

In fact, we should probably be specific about both cases:

```rust
(Some(_), Some(_)) => unreachable!("TimeZoneFormat should have a configuration or a pattern, but found both"),
(None, None) => unreachable!("TimeZoneFormat should have either a configuration or a pattern, but found neither"),
```

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/time_zone.rs:174]: > nitpick

Same here with the grouping of `config` and `patterns` (though it makes no difference to the compiler).

```rust
let mut time_zone_format = Self {
    config: None,
    patterns: Some(patterns),
    zone_formats,
    exemplar_cities: None,
    mz_generic_long: None,
    mz_generic_short: None,
    mz_specific_long: None,
    mz_specific_short: None,
};
```

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/time_zone.rs:113]: > nitpick

I think it would make more sense to readers to move `config` to the top where `patterns` is, since those are the two members for initializing the TZF, and then the rest of the members are data payloads.

**Review by nordzilla — CHANGES_REQUESTED** [review]
Just a couple small things I missed on the first pass, otherwise this looks really great to me! 

I'll leave it to Shane to lend a second pair of eyes before we land this. 

Really excited to make this public!

**Comment by samchen61661** [comment]
> Just a couple small things I missed on the first pass, otherwise this looks really great to me!
> 
> I'll leave it to Shane to lend a second pair of eyes before we land this.
> 
> Really excited to make this public!

Thanks for your guidance Eric! I learnt a lot from you these days! I did not expect that it touched so many files eventually and I felt that I am more comfortable to read and write Rust code. And I am very happy to make `TimeZoneConfig` public!

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [reply components/datetime/src/time_zone.rs:825]: @samchen61661 

Regardless of whether this can be ignored by DCE, it's failing our CI tests for building with/without all features enabled, because `serde` is not always available. 

So I believe we will have to do what I suggested above:

We need to remove the `Serialize` and `Deserialize` dependencies from `TimeZoneFormatConfig`, `IsoFormat`, `IsoMinutes`, `IsoSeconds`, and we need to make essentially the same types where we define the other (de)serializable sttructs for the test fixtures:

https://github.com/unicode-org/icu4x/blob/main/components/datetime/tests/patterns/structs/time_zones.rs

Then we need to implement the [`From`](https://doc.rust-lang.org/std/convert/trait.From.html) trait from the (de)serializable test types to turn into the real types in this file same.

So in `tests/patterns/structs/time_zones.rs` you would have something like this:

```rust
use icu_datetime::time_zone;

...

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub enum IsoSeconds {
    Optional,
    Never,
}

impl From<IsoSeconds> for time_zone::IsoSeconds {
    fn from(other: IsoSeconds) -> time_zone::IsoSeconds {
        match other {
            IsoSeconds::Optional => time_zone::IsoSeconds::Optional,
            IsoSeconds::Never => time_zone::IsoSeconds::Never,
        }
    }
}
```

Then do this for all of the enums, including `TimeZoneFormatConfig`

Implementing the `From` trait enables the `.into()` functionality as well, so when it comes time to do the `Iso8601` variant itself, you should be able to just do something like:

```rust
match other {
    ...
    Iso8601(iso_format, iso_minutes, iso_seconds) => time_zone::TimeZoneFormatConfig::Iso8601(
        iso_format.into(),
        iso_minutes.into(),
        iso_seconds.into(),
    ),
}
```

**Review by nordzilla — APPROVED** [review]
Thanks so much for working on this. This looks good to me. 

Technically it's missing some documentation comments, such as on the newly added function `TimeZoneFormat::try_from_config()`, but once this lands I'm planning to go through and read all the docs/update them/add anything that's missing, so I'm comfortable landing this as is with docs as a follow-up.

File: components/datetime/src/format/time_zone.rs
Comment (sffc) [inline components/datetime/src/format/time_zone.rs:69]: Suggestion: To avoid the need for the `.expect()` a few lines down, you could add an argument `config: &TimeZoneFormatConfig` which you populate at the call site where you have the variable.

File: components/datetime/src/format/time_zone.rs
Comment (sffc) [inline components/datetime/src/format/time_zone.rs:115]: Suggestion: same as above but for patterns.

File: components/datetime/src/time_zone.rs
Comment (sffc) [inline components/datetime/src/time_zone.rs:67]: Question (here and below): Is it time to uncomment these tests?

File: components/datetime/src/time_zone.rs
Comment (sffc) [inline components/datetime/src/time_zone.rs:179]: Suggestion: if you move the `let zone_symbols = ...` above the `let mut time_zone_format = ...` then you don't have to do this because you haven't handed off ownership yet.

File: components/datetime/src/time_zone.rs
Comment (sffc) [inline components/datetime/src/time_zone.rs:315]: Suggestion: the config enum can be `Copy` and you can just copy it so you don't have to do the `expect()`.

**Review by sffc — CHANGES_REQUESTED** [review]
Looks good overall.  I found some small areas for improvement.  We'll probably need to add more configs to this enum in the future, but now we have a straightforward place to add them.

For example:
https://tc39.es/proposal-intl-extend-timezonename/#sec-basicformatmatcher

File: components/datetime/src/time_zone.rs
Comment (nordzilla) [reply components/datetime/src/time_zone.rs:160]: @sffc 

I'm not quite clear on what you're suggesting here. 

Do you want us to clone the `patterns` argument?

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/format/time_zone.rs:70]: > nitpick

Since this is a `Copy` type we should just take it by value, instead of by reference.

File: components/datetime/src/format/time_zone.rs
Comment (nordzilla) [inline components/datetime/src/format/time_zone.rs:63]: @sffc 

By your same logic about the `expect` potentially adding unwanted panic machinery, should we remove the `unreachable!()` calls here and return an error instead?

**Review by nordzilla — CHANGES_REQUESTED** [review]
This looks good. Just one small nitpick from me.

Good call on removing those `expect` calls @sffc.

File: components/datetime/src/format/time_zone.rs
Comment (samchen61661) [reply components/datetime/src/format/time_zone.rs:70]: I was using `TimeZoneFormatConfig` at the beginning, but it turned out that config is type `&TimeZoneFormatConfig`. If we want to use `config: TimeZoneFormatConfig`, should I use `(None, Some(config)) => write_config(time_zone_format, time_zone, *config, w)` in `write_zone` method above?

Comment (nordzilla) [reply components/datetime/src/format/time_zone.rs:70]: @samchen61661 

Just remove the reference from the top-level match

```diff
-    match (&time_zone_format.patterns, &time_zone_format.config) {
+    match (&time_zone_format.patterns, time_zone_format.config) {
```

And then you should be able to take everything else by value, including `IsoMinutes` etc. no problem.

Comment (samchen61661) [reply components/datetime/src/format/time_zone.rs]: Thanks Eric! It works!

**Review by nordzilla — DISMISSED** [review]
This looks good to me, just pending my one question to Shane regarding the panic handling:
https://github.com/unicode-org/icu4x/pull/1256#discussion_r750785392

File: components/datetime/src/time_zone.rs
Comment (sffc) [inline components/datetime/src/time_zone.rs:64]: Nit: Update the prose above this to update references to "patterns" with the new config enum

**Review by sffc — DISMISSED** [review]

File: components/datetime/src/format/time_zone.rs
Comment (sffc) [reply components/datetime/src/format/time_zone.rs:63]: I'm not quite as concerned about a plain `unreachable!()` because it's more self-contained; `.expect()` pulls in the formatting machinery since it prints a formatted message\*.  If we build with `panic_abort`, then any `unreachable!()` should basically compile away.

That being said, if you want to avoid the unreachable code path, you can use an internal enum such as

```rust
enum TimeZoneFormatKind {
  Config(TimeZoneFormatConfig),
  Pattern(DataPayload<'data, PatternPluralsFromPatternsV1Marker>),
}
```

This would likely be a good cleanup to your code regardless.

**Review by nordzilla — APPROVED** [review]

File: components/datetime/src/format/time_zone.rs
Comment (sffc) [reply components/datetime/src/format/time_zone.rs:63]: I'm re-opening this thread because I think switching to use an enum may be a beneficial change for @samchen61661 to consider.

Comment (samchen61661) [reply components/datetime/src/format/time_zone.rs:63]: Yeah, it is a good suggestion. I am not familiar with the open source policy. What do you think to improve it in a separate PR?

The main concern here is that this PR is big and the fix is relatively small.

Comment (nordzilla) [reply components/datetime/src/format/time_zone.rs:63]: Got it. 

When you said "This would likely be a good cleanup to your code regardless." I thought you meant as a follow-up. 

Happy to add it in here to as I agree that it is a good change. 

@samchen61661 you can go ahead and add it to this PR.

**Review by sffc — APPROVED** [review]

File: components/datetime/src/format/time_zone.rs
Comment (sffc) [reply components/datetime/src/format/time_zone.rs:63]: I'm fine landing this as-is and doing this refactor in a small follow-up PR.  But since you still need to resolve the merge conflict in this PR, it seems to make sense to fix this while you're at it.

Comment (samchen61661) [reply components/datetime/src/format/time_zone.rs:63]: I see. sg and I will refactor it in this PR.

Comment (samchen61661) [reply components/datetime/src/format/time_zone.rs:63]: Shane, when you talked about `you can use an INTERNAL exhaustive enum`, are you saying that I just need to use `TimeZoneFormatKind` in the `write_zone` method?

I think we can make `TimeZoneFormatKind` as an external attribute in `TimeZoneFormat` to replace `config` and `patterns`.

Comment (nordzilla) [reply components/datetime/src/format/time_zone.rs:63]: @samchen61661 

What we would need to do is:

```rust
pub(super) TimeZoneFormatKind<'data> {
    ...
}
```

The enum should only be internal (i.e. `pub(super)`) because users of the public API shouldn't have to worry about this type. 

And then instead of having data members for `config` and `patterns`, you would just have one data member `kind` that is a `TimeZoneFormatKind`.

Note that the enum `TimeZoneFormatKind` will need to have a lifetime specifier since it holds the `DataPayload` for the pattern variant.

**Review by nordzilla — APPROVED** [review]
Looks good to me!

Thanks for being willing to make all of these small changes over several iterations.

I think the code is overall cleaner because of it.

**Review by sffc — APPROVED** [review]

**Pull request merged**
